maintainers/scripts/update-typst-packages.py

```diff
@@ -0,0 +1,213 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -p "python3.withPackages (p: with p; [ tomli packaging ])" nixfmt-rfc-style -i python3
+
+import os, re, tomli, subprocess, concurrent.futures
+import argparse, tempfile, tarfile
+from string import Template, punctuation
+from packaging.version import Version
+from urllib import request
+
+class TypstPackage:
+    def __init__(self, **kwargs):
+        self.pname = kwargs["pname"]
+        self.version = kwargs["version"]
+        self.name_nix = self.package_version_to_nix(self.pname, self.version)
+        self.meta = kwargs["meta"]
+        self.path = kwargs["path"]
+
+        self.repo = \
+            None if "repository" not in self.meta["package"] else self.meta["package"]["repository"]
+        self.description = self.meta["package"]["description"].rstrip(punctuation)
+        self.license = self.meta["package"]["license"]
+        self.params = "" if "params" not in kwargs else kwargs["params"]
+        self.deps = [] if "deps" not in kwargs else kwargs["deps"]
+        self.template = Template("""\
+{
+  lib,
+  buildTypstPackage,$maybeparams
+}:
+buildTypstPackage rec {
+  pname = "$pname";
+  version = "$version";
+
+  src = $src;
+
+  sourceRoot = ".";$maybedeps
+
+  meta = {$maybehomepage
+    description = "$description";
+    license = $license;
+    maintainers = with lib.maintainers; [ cherrypiejam ];
+  };
+}\
+""")
+
+    @classmethod
+    def license_string_to_nix(cls, license_string):
+        license_tokens = re.split(r" AND | OR ", license_string)
+        license_nix = "["
+        for token in license_tokens:
+            license_nix += f" (lib.getLicenseFromSpdxId \"{token}\")"
+        return license_nix + " ]"
+
+    @classmethod
+    def package_version_to_nix(cls, package, version):
+        package_token_head, *package_token_tail = package.split('-')
+        package_nix = ''.join([package_token_head] + list(map(lambda t: t.title(), package_token_tail)))
+        version_number = map(lambda x: int(x), version.split('.'))
+        version_nix = '_'.join(map(lambda x: str(x), version_number))
+        return '_'.join((package_nix, version_nix))
+
+    @classmethod
+    def deps_to_nix(cls, deps):
+        if len(deps) == 0:
+            return None
+        block = "with typstPackages; ["
+        for p, v in deps:
+            block += " " + cls.package_version_to_nix(p, v)
+        block += " ]"
+        return block
+
+    @classmethod
+    def source_to_nix(cls, package, version):
+        cmd_template = "nix-build --no-out-link -E '{}'"
+        cmd = [ "nix-build", "--no-out-link", "-E" ]
+        nix_template = Template("""\
+builtins.fetchurl {
+    url = "$url";
+    sha256 = "$hash";
+  }""")
```

Comment:
Note that `builtins.fetchurl` is unavailable under restricted evaluation mode.

Reply:
I looked the docs and I think it is fine here. The doc says under the restricted eval mode, nix will not access any file outside of `nixPath`, which we have to, in our case, bc we are fetching packages from Typst Universe, no matter what fetchers we are using.

Reply:
> I looked the docs and I think it is fine here. The doc says under the restricted eval mode, nix will not access any file outside of `nixPath`, which we have to, in our case, bc we are fetching packages from Typst Universe, no matter what fetchers we are using.

I agree with you here.

However, we should still use a fetcher that produces a derivation here instead of a built-in fetcher. The built-in fetcher fetches during evaluation time, which is non-parallel, slow to evaluate, and causes evaluation failure without network connections.

Reply:
That tracks. Thank for pointing it out. Didn't know that